### PR TITLE
Adding todo config util

### PR DESCRIPTION
**Pull request by scalvert**
Adds new util to retrieve the todo configuration. 

```ts
export function getTodoConfig(
  baseDir: string,
  todoConfig: TodoConfig = {}
): TodoConfig | undefined;
```

Config values can be present in

1. The package.json

    ```
    {
		"lintTodo": {
			"daysToDecay": {
				"warn": 5,
				"error": 10
			}
		}
	}
    ```
1. Environment variables (`TODO_DAYS_TO_WARN` or `TODO_DAYS_TO_ERROR`)
	- Env vars override package.json config
1. Passed in directly, such as from command line options.
	- Passed in options override both env vars and package.json config

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,4 @@
 node_modules
 lib
 tsconfig.tsbuildinfo
+.vscode/
```
Comment (rwjblue) [inline .gitignore:4]: 😭 should be in your global .gitconfig (avoids having to change every project for editor specific settings)

https://docs.github.com/en/free-pro-team@latest/github/using-git/ignoring-files#configuring-ignored-files-for-all-repositories-on-your-computer

File: README.md [hunk 2]
```diff
@@ -340,5 +347,36 @@ Applies todo changes, either adding or removing, based on batches from `getTodoB
...
+| todoConfig | The optional todo configuration. |
+
+**Example**  
+```
+{
+  "lintTodo": {
+    "daysToDecay": {
+      "warn": 5,
+			 "error": 10
+		 }
```
Comment (rwjblue) [inline README.md:373]: Seems wrong, I think there is an extra `}` and the `"error"` line is indented differently than the others

File: __tests__/get-todo-config-test.ts [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+  let project: FakeProject;
+
+  beforeEach(() => {
+    project = new FakeProject();
+    project.writeSync();
+  });
+
+  afterEach(() => {
+    project.dispose();
+  });
```
Comment (rwjblue) [inline __tests__/get-todo-config-test.ts:14]: This needs to account for tearing down the various "setupEnvVar"s (or tests bleed into each other)

File: src/builders.ts [hunk 4]
```diff
@@ -14,7 +14,7 @@ import { DaysToDecay, FilePath, LintMessage, LintResult, TodoData } from './type
 export function buildTodoData(
   baseDir: string,
   lintResults: LintResult[],
-  daysToDecay?: DaysToDecay
+  daysToDecay?: TodoConfig
```
Comment (rwjblue) [inline src/builders.ts:17]: Any objection to changing the variable name also?

File: src/builders.ts [hunk 5]
```diff
@@ -47,7 +47,7 @@ export function _buildTodoDatum(
   baseDir: string,
   lintResult: LintResult,
   lintMessage: LintMessage,
-  daysToDecay?: DaysToDecay
+  daysToDecay?: TodoConfig
```
Comment (rwjblue) [inline src/builders.ts:50]: Can we make the variable match the updated interface name?

File: src/get-todo-config.ts [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+ *
+ * 1. The package.json
+ * @example
+ * ```
+ * {
+ *   "lintTodo": {
+ *     "daysToDecay": {
+ *       "warn": 5,
+ * 			 "error": 10
+ * 		 }
```
Comment (rwjblue) [inline src/get-todo-config.ts:16]:
```suggestion
 *       "warn": 5,
 *       "error": 10
 * 	   }
```

File: tsconfig.json [hunk 7]
```diff
@@ -2,10 +2,12 @@
   "compilerOptions": {
     "declaration": true,
     "declarationMap": true,
+    "esModuleInterop": true,
     "sourceMap": true,
     "importHelpers": true,
     "module": "commonjs",
     "strict": true,
+    "skipLibCheck": true,
```
Comment (rwjblue) [inline tsconfig.json:10]: Why do we need this?

**Review by rwjblue — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 1 above)
Comment (scalvert) [reply .gitignore:4]: Uggg yes. Will do that now. Sorry. If I had a penny for every time you've said that to me...

File: README.md (hunk 2 above)
Comment (scalvert) [reply README.md:373]: Ya the auto generator borked the indentation.

File: __tests__/get-todo-config-test.ts (hunk 3 above)
Comment (scalvert) [reply __tests__/get-todo-config-test.ts:14]: Each test is tearing down its own env vars, so there's no bleeding, unless you see one that's not correctly tearing things down.

File: src/builders.ts (hunk 4 above)
Comment (scalvert) [reply src/builders.ts:17]: Oh no, this was just a miss on my part.

File: README.md (hunk 2 above)
Comment (scalvert) [reply README.md:373]: Also it was actually _missing_ a `}`

File: src/builders.ts (hunk 5 above)
Comment (scalvert) [reply src/builders.ts:50]: Just a miss on my part.

File: src/get-todo-config.ts [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+function getFromPackageJson(basePath: string): TodoConfig {
+  let pkg;
+  const packageJsonPath = join(basePath, 'package.json');
+
+  try {
+    pkg = require(packageJsonPath);
+  } catch (error) {
+    if (error.code === 'ENOENT') {
+      throw new Error(`The ${basePath} directory does not contain a package.json file.`);
+    }
```
[the remark this answers is not in the document]
Comment (scalvert) [reply src/get-todo-config.ts:59]: Ya that makes sense.

File: src/get-todo-config.ts [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+
+  if (Number.isInteger(error)) {
+    config.error = error;
+  }
+
+  return config;
+}
+
+function getEnvVar(name: string) {
+  if (typeof process.env[name] !== 'undefined' && process.env[name] !== '') {
```
[the remark this answers is not in the document]
Comment (scalvert) [reply src/get-todo-config.ts:83]: Totally could. I find with TS I tend to go overboard on type checking. I'll totally update it.

File: src/io.ts [hunk 10]
```diff
@@ -121,19 +108,32 @@ export function writeTodosSync(
...
+ * @param daysToDecay - An object containing the warn or error days, in integers.
+ * @returns - The todo storage directory path.
+ */
 export function writeTodosSync(
   baseDir: string,
   lintResults: LintResult[],
-  filePath?: string | DaysToDecay,
-  daysToDecay?: DaysToDecay
+  filePath?: string | TodoConfig,
+  daysToDecay?: TodoConfig
```
[the remark this answers is not in the document]
Comment (scalvert) [reply src/io.ts:136]: Yep!

File: tsconfig.json [hunk 11]
```diff
@@ -2,10 +2,12 @@
   "compilerOptions": {
     "declaration": true,
     "declarationMap": true,
+    "esModuleInterop": true,
```
[the remark this answers is not in the document]
Comment (scalvert) [reply tsconfig.json:5]: Fair point. Will swap it over and see if that fixes the original issue.

File: tsconfig.json (hunk 7 above)
Comment (scalvert) [reply tsconfig.json:10]: We don't.

File: .gitignore (hunk 1 above)
Comment (scalvert) [reply .gitignore:4]: Done!

File: README.md (hunk 2 above)
Comment (scalvert) [reply README.md:373]: Fixed.

File: src/builders.ts (hunk 4 above)
Comment (scalvert) [reply src/builders.ts:17]: Fixed.

File: src/builders.ts (hunk 5 above)
Comment (scalvert) [reply src/builders.ts:50]: Fixed.

File: src/get-todo-config.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (scalvert) [reply src/get-todo-config.ts:83]: Done!

File: src/io.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (scalvert) [reply src/io.ts:136]: Done!

File: tsconfig.json (hunk 11 above)
[the remark this answers is not in the document]
Comment (scalvert) [reply tsconfig.json:5]: No dice. Error:

Screen Shot 2021-01-05 at 2 16 32 PM

File: tsconfig.json (hunk 7 above)
Comment (scalvert) [reply tsconfig.json:10]: Removed.

File: __tests__/get-todo-config-test.ts (hunk 3 above)
Comment (scalvert) [reply __tests__/get-todo-config-test.ts:16]: I added an explicit teardown util so it's more clear that's what's going on.

File: src/get-todo-config.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (scalvert) [reply src/get-todo-config.ts]: Done.

File: __tests__/get-todo-config-test.ts [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+
+  beforeEach(() => {
+    project = new FakeProject();
+    project.writeSync();
+  });
+
+  afterEach(() => {
+    project.dispose();
+    delete process.env['TODO_DAYS_TO_WARN'];
+    delete process.env['TODO_DAYS_TO_ERROR'];
```
Comment (rwjblue) [inline __tests__/get-todo-config-test.ts:17]: Probably don't need to brackets + quotes. Not a big deal though.

File: src/get-todo-config.ts [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+ *
+ * The package.json
+ *
+ * @example
+ * ```json
+ * {
+ *   "lintTodo": {
+ *     "daysToDecay": {
+ *       "warn": 5,
+ * 			 "error": 10
```
Comment (rwjblue) [inline src/get-todo-config.ts:16]: This indentation seems wrong still

File: src/get-todo-config.ts [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+      'The `lintTodo` configuration in the package.json contains invalid values. The `warn` value must be less than the `error` value.'
+    );
+  }
+
+  return mergedConfig;
+}
+
+function getFromPackageJson(basePath: string): TodoConfig {
+  // eslint-disable-next-line @typescript-eslint/no-var-requires
+  const pkg = require(join(basePath, 'package.json'));
```
Comment (rwjblue) [inline src/get-todo-config.ts:55]: This will error if there is no package.json (which we don't want)

**Review by rwjblue — CHANGES_REQUESTED** [review]

File: __tests__/get-todo-config-test.ts (hunk 12 above)
Comment (scalvert) [reply __tests__/get-todo-config-test.ts:17]: Ya, I hate seeing snake cased props, so it 'feels' better to me to be strings. 🤷‍♂️

File: src/get-todo-config.ts (hunk 13 above)
Comment (scalvert) [reply src/get-todo-config.ts:16]: I think I'm being trolled by vscode...

Comment (scalvert) [reply src/get-todo-config.ts]: Fixed. Had some tabs in there.

File: src/get-todo-config.ts (hunk 14 above)
Comment (scalvert) [reply src/get-todo-config.ts]: Fixed.

**Review by rwjblue — APPROVED** [review]

**Pull request merged**